---
title: "HaploDup report"
output:
  flexdashboard::flex_dashboard:
    orientation: columns
    vertical_layout: fill
    self_contained: true
params:
  coords: ""
  coords_self: ""
  counts_hap1: ""
  counts_hap2: ""
  min_align: ""
  similarity: ""
  queryID: ""
  refID: ""
  hap1ID: ""
  hap2ID: ""
  hap1Len: ""
  hap2Len: ""
  ratio: ""
  structure : ""
  legacy : ""
  markers : ""
  dup_markers : ""
---


```{r}
# load libraries
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(plotly))
suppressPackageStartupMessages(library(htmlwidgets))
suppressPackageStartupMessages(library(RColorBrewer))


# prepare variables
nucmer_file <- params$coords
nucmer_file_self <- params$coords_self
nucmer_query <- paste0(params$queryID)
nucmer_ref <- paste0(params$refID)
nucmer_title <- paste0(params$queryID," on " , params$refID)

counts_hap1_file <- params$counts_hap1
counts_hap2_file <- params$counts_hap2
gmap_hap1 <- params$hap1ID
gmap_hap2 <- params$hap2ID
hap1Len <- as.numeric(params$hap1Len)
hap2Len <- as.numeric(params$hap2Len)

min_len <- as.numeric(params$min_align)
min_iden <- as.numeric(params$similarity)
ratio <- as.numeric(params$ratio)

structure_file <- params$structure
legacy_file <- params$legacy
markers_all_file <- params$markers
markers_dup_file <- params$dup_markers

category_conversion_1 <- c(
"dup_1_0" = "CDS in sinlge copy in Hap1 but with no hit in Hap2" ,
"dup_gt1_0" = "CDS with multiple hits in Hap1 but no hit in Hap2" ,
"dup_gt1_gt0_grt" = paste0("CDS present in both haplotypes, Hap1/Hap2 counts ratio > ",(1/ratio)) ,
"dup_gt1_gt0_lrt" = paste0("CDS present in both haplotypes, Hap1/Hap2 counts ratio < ",ratio)
)

category_conversion_2 <- c(
"dup_0_1" = "CDS in sinlge copy in Hap2 but no hit in Hap1" ,
"dup_0_gt1" = "CDS with multiple hits in Hap2 but no hit in Hap1" ,
"dup_gt0_gt1_grt" = paste0("CDS present in both haplotypes, Hap2/Hap1 counts ratio > ",(1/ratio)) ,
"dup_gt0_gt1_lrt" = paste0("CDS present in both haplotypes, Hap2/Hap1 counts ratio < ",ratio)
)

# load data
tb_1 <- read_delim(file = nucmer_file,
                 delim = "\t",
                 col_names = T)

# Columns:
# tID
# tLen
# tStart
# tStop
# qID
# qLen
# qStart
# qStop
# identity
# match

# TODO: Add self map of reference on p0 plot

diploid_trace_hap1 <- read_delim(counts_hap1_file,
                            delim = "\t",
                            col_names = T)


diploid_trace_hap2 <- read_delim(counts_hap2_file,
                            delim = "\t",
                            col_names = T)

# Columns:
# Chr
# Start
# Stop
# Gene_Id
# Hap1_count
# Hap2_count
# Hap2_to_Hap1_ratio
# Description
# Gene_count_with_description


structure_trace <- read_delim(structure_file,
                            delim = "\t",
                            col_names = c( "Chr" , "Start" , "Stop" , "Seq_Id" , "orientation" ) ,
                            )

legacy_trace <- read_delim(legacy_file,
                            delim = "\t",
                            col_names = c( "Chr" , "Start" , "Stop" , "Seq_Id" , "orientation" ) ,
                            )
# subset
coords1 <- tb_1 %>% filter(tID==nucmer_ref , qID==nucmer_query , identity > min_iden , match > min_len )
refLen <- max(coords1$tLen)
queryLen <- max(coords1$qLen)

diploid_trace_ref <- bind_rows(diploid_trace_hap1,diploid_trace_hap2) %>% filter(Chr==nucmer_ref)
diploid_trace_ref <- pivot_longer(diploid_trace_ref, cols = c("Hap1_count", "Hap2_count"), names_to = "HAP", values_to = "count")

diploid_trace_hap1 <- diploid_trace_hap1 %>%
  mutate(test = Hap1_count/Hap2_count) %>%
  mutate(dup_1_0 = (Hap1_count==1 & Hap2_count==0),
         dup_gt1_0 = (Hap1_count>1 & Hap2_count==0),
         dup_gt1_gt0_grt = (Hap1_count>1 & Hap2_count>0 & test > (1/ratio) ) ,
         dup_gt1_gt0_lrt = (Hap1_count>1 & Hap2_count>0 & test < ratio ) ) %>%
  filter(Chr==gmap_hap1)

diploid_trace11 <- pivot_longer(diploid_trace_hap1, cols = c("Hap1_count", "Hap2_count"), names_to = "HAP", values_to = "count")
diploid_trace12 <- pivot_longer(diploid_trace_hap1, cols = c("dup_1_0", "dup_gt1_0", "dup_gt1_gt0_grt" , "dup_gt1_gt0_lrt" ), names_to = "DUP", values_to = "DUP_test")
diploid_trace12 <- diploid_trace12 %>%
  mutate(DUP_test = str_replace_all(DUP_test, "FALSE", "0")) %>%
  mutate(DUP_test = str_replace_all(DUP_test, "TRUE", "1")) %>%
  mutate(DUP_test = as.numeric(DUP_test))

#- dup_1_0: CDS of hap1 absent in hap2
#- dup_gt1_0: CDS present more than once in hap1 but absent in hap2
#- dup_gt1_gt0_lrt: CDS present more than once in hap1, at least once in hap2, with a ratio hap1/hap2 greater than given ratio


diploid_trace_hap2 <- diploid_trace_hap2 %>%
  mutate(test = Hap2_count/Hap1_count) %>%
  mutate(dup_0_1 = (Hap1_count==0 & Hap2_count==1),
         dup_0_gt1 = (Hap1_count==0 & Hap2_count>1),
         dup_gt0_gt1_grt = (Hap1_count>0 & Hap2_count>1 & test > (1/ratio) ) ,
         dup_gt0_gt1_lrt = (Hap1_count>0 & Hap2_count>1 & test < ratio ) ) %>%
  filter(Chr==gmap_hap2)

diploid_trace21 <- pivot_longer(diploid_trace_hap2, cols = c("Hap1_count", "Hap2_count"), names_to = "HAP", values_to = "count")
diploid_trace22 <- pivot_longer(diploid_trace_hap2, cols = c("dup_0_1", "dup_0_gt1", "dup_gt0_gt1_grt" , "dup_gt0_gt1_lrt"), names_to = "DUP", values_to = "DUP_test")
diploid_trace22 <- diploid_trace22 %>%
  mutate(DUP_test = str_replace_all(DUP_test, "FALSE", "0")) %>%
  mutate(DUP_test = str_replace_all(DUP_test, "TRUE", "1")) %>%
  mutate(DUP_test = as.numeric(DUP_test))

#- dup_0_1: CDS of hap2 absent in hap1
#- dup_0_gt1: CDS present more than once in hap2 but absent in hap1
#- dup_gt0_gt1_lrt: CDS present more than once in hap2, at least once in hap1, with a ratio hap2/hap1 greater than 2



all_marker_trace_r <- all_marker_trace %>% filter(Chr==nucmer_ref )
all_marker_trace_h1 <- all_marker_trace %>% filter(Chr==gmap_hap1 )
all_marker_trace_h2 <- all_marker_trace %>% filter(Chr==gmap_hap2 )
dup_markers_trace_r <- dup_markers_trace %>% filter(Chr==nucmer_ref )
dup_markers_trace_h1 <- dup_markers_trace %>% filter(Chr==gmap_hap1 )
dup_markers_trace_h2 <- dup_markers_trace %>% filter(Chr==gmap_hap2 )

structure_trace_r <- structure_trace %>% filter(Chr==nucmer_ref )
structure_trace_h1 <- structure_trace %>% filter(Chr==gmap_hap1 )
structure_trace_h2 <- structure_trace %>% filter(Chr==gmap_hap2 )
legacy_trace_r <- legacy_trace %>% filter(Chr==nucmer_ref )
legacy_trace_h1 <- legacy_trace %>% filter(Chr==gmap_hap1 )
legacy_trace_h2 <- legacy_trace %>% filter(Chr==gmap_hap2 )


a <- list(
  text = "Gene counts",
  xref = "paper",
  yref = "paper",
  yanchor = "bottom",
  x = 0,
  y = 1,
  showarrow = FALSE
)

b <- list(
  text = "",
  xref = "paper",
  yref = "paper",
  yanchor = "bottom",
  x = 0,
  y = 1,
  showarrow = FALSE
)

c <- list(
  text = "Markers",
  xref = "paper",
  yref = "paper",
  yanchor = "bottom",
  x = 0,
  y = 1,
  showarrow = FALSE
)

d <- list(
  text = "Duplicated Markers",
  xref = "paper",
  yref = "paper",
  yanchor = "bottom",
  x = 0,
  y = 1,
  showarrow = FALSE
)


e <- list(
  text = "Input sequence composition",
  xref = "paper",
  yref = "paper",
  yanchor = "bottom",
  x = 0,
  y = 1,
  showarrow = FALSE
)

f <- list(
  text = "Legacy sequence composition",
  xref = "paper",
  yref = "paper",
  yanchor = "bottom",
  x = 0,
  y = 1,
  showarrow = FALSE
)

```


```{r}
# plot
p0 <- ggplot() +
  geom_segment( data = coords1 ,
    aes( x = tStart,
         y = qStart,
         xend = tStop,
         yend = qStop,
         colour = nucmer_query ,
         text = sprintf(
        'Query ID: %s<br>Target ID: %s<br>Qstart: %s - Qend: %s<br>Tstart: %s - Tend: %s<br>Identity: %s',
        qID, tID, qStart, qStop, tStart, tStop , identity
          )
        )
  ) +
#  geom_point(aes(x = tStart, y = qStart, color = identity), size = 0.009) +
#  geom_point(aes(x = tStop, y = qStop, color = identity),size = 0.009) +
  scale_color_manual(name = "Sequence", breaks = c( nucmer_query , paste0( nucmer_ref , "<br>(Self)") ) , values = c("black", "gray66") ) +
  scale_x_continuous(limits = c(0, refLen)) +
  scale_y_continuous(limits = c(0, max(refLen , queryLen) )) +
  labs(title = nucmer_title, x = nucmer_ref, y = "Query sequence - Position (bp)")

p0 <- p0 + theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
               axis.title.x = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
               axis.title.y = element_text(color="black", size=10, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm")),
               axis.text.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
               axis.text.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
               axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
               axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
               axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
               axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
               plot.background = element_blank(),
               panel.background = element_blank(),
               panel.grid.major.y = element_line(color="gray80", size=0.2, linetype="solid") ,
               panel.grid.major.x = element_line(color="gray80", size=0.2, linetype="solid"),
               panel.grid.minor.y = element_blank(),
               panel.grid.minor.x = element_blank(),
               plot.margin=unit(c(1,1,1,1),"mm"),
               legend.position="right")

pl0 <- ggplotly( p0, tooltip = "text")

```


```{r}

p1 <- ggplot(diploid_trace11,
    aes(
      x = Start,
      y = count,
      color = HAP,
      group = 1,
      text = sprintf(
        'Gene ID: %s<br>Gene start: %s<br>%s: %s',
        Gene_Id, Start , HAP , count
          )
    )
  ) +
  geom_step() +
  scale_colour_manual(values = brewer.pal(8, "Dark2")[1:2] , labels = c("Hap 1 counts" , "Hap 1 counts") ) +
  scale_y_continuous(breaks = seq(min(diploid_trace11$count), max(diploid_trace11$count), 1)) +
  labs(x = "Sequence position (bp)", y = "Gene alignments counts" , color = "")

p1 <- p1 +
          theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
                 axis.title.x = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.title.y = element_text(color="black", size=10, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm")),
                 axis.text.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
                 axis.text.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
                 axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
                 plot.background = element_blank(),
                 panel.background = element_blank(),
                 panel.grid.major.y = element_blank(),
                 panel.grid.major.x = element_blank(),
                 panel.grid.minor.y = element_blank(),
                 panel.grid.minor.x = element_blank(),
                 plot.margin=unit(c(.7,.7,.7,.7),"mm"),
                 legend.position="right"
            )

pl1 <- ggplotly(p1, tooltip = "text") %>% layout(annotations = a , legend=list(title="") , xaxis = list(title="Sequence position (bp)") )

```

```{r}
#- dup_1_0: CDS of hap1 absent in hap2
#- dup_gt1_0: CDS present more than once in hap1 but absent in hap2
#- dup_gt1_gt0_rtgt2: CDS present more than once in hap1, at least once in hap2, with a ratio hap1/hap2 greater than 2

p2 <- ggplot(diploid_trace12 ,
    aes(
    x = Start,
    y = DUP_test,
    color = DUP,
    group = 1,
    text = sprintf(
        'Gene ID: %s<br>Gene start: %s<br>Test result: %s',
        Gene_Id, Start , DUP_test
          )
    )
  ) +
  geom_step() +
  facet_wrap(~DUP , ncol=1 , scales = "free_y", labeller = as_labeller(category_conversion_1) ) +
  scale_y_continuous(breaks = c(0,1) , limits=c(0,1.1) , labels = c("False" , "True")) +
  scale_colour_manual(values = brewer.pal(8, "Dark2")[3:6] ) +
  labs(x = "Sequence position (bp)", y = "Test result" , color = "")

p2 <- p2 +
        theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
                 axis.title.x = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.title.y = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.text.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
                 axis.text.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
                 axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
                 plot.background = element_blank(),
                 panel.background = element_rect(fill = "white", size = 0.5, linetype = "solid"),
                 panel.grid.major.y = element_line(color="#777777", size=0.2, linetype="solid"),
                 panel.grid.major.x = element_blank(),
                 panel.grid.minor.y = element_blank(),
                 panel.grid.minor.x = element_blank(),
                 plot.margin=unit(c(.7,.7,.7,.7),"mm"),
                 legend.position="right",
                 panel.spacing = unit(.8, "lines") ,
                 strip.background = element_rect(fill = "#999999" ),
                 strip.text.x = element_text(margin = margin(.3, 0, .1, 0, "lines") , color = "white ")
    )

pl2 <- ggplotly(p2, tooltip = "text") %>% layout(annotations = b , legend=list(title="") , xaxis = list(title=list( text="Sequence position (bp)", font = list(size=13) ) ) )

```


```{r}

p3 <- ggplot(diploid_trace21,
  aes(
    x = Start,
    y = count,
      color = HAP,
      group = 1,
      text = sprintf(
        'Gene ID: %s<br>Gene start: %s<br>%s: %s',
        Gene_Id, Start , HAP , count
          )
    )
  ) +
  geom_step() +
  scale_colour_manual(values = brewer.pal(8, "Dark2")[1:2]) +
  scale_y_continuous(breaks = seq(min(diploid_trace21$count), max(diploid_trace21$count), 1)) +
  labs(x = "Sequence position (bp)", y = "Gene alignments counts" , color = "")

p3 <- p3 +
        theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
                 axis.title.x = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.title.y = element_text(color="black", size=10, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm")),
                 axis.text.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
                 axis.text.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
                 axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
                 plot.background = element_blank(),
                 panel.background = element_blank(),
                 panel.grid.major.y = element_blank(),
                 panel.grid.major.x = element_blank(),
                 panel.grid.minor.y = element_blank(),
                 panel.grid.minor.x = element_blank(),
                 plot.margin=unit(c(.7,.7,.7,.7),"mm"),
                 legend.position="right"
          )

pl3 <- ggplotly(p3, tooltip = "text") %>% layout(annotations = a , legend=list(title="") , xaxis = list(title="Sequence position (bp)") )

```

```{r}
#- dup_0_1: CDS of hap2 absent in hap1
#- dup_0_gt1: CDS present more than once in hap2 but absent in hap1
#- dup_gt0_gt1_lrt: CDS present more than once in hap2, at least once in hap1, with a ratio hap2/hap1 greater than 2


p4 <- ggplot(diploid_trace22) +
  geom_step(
    aes(
    x = Start,
    y = DUP_test,
    color = DUP,
    group=1,
    text = sprintf(
        'Gene ID: %s<br>Gene start: %s<br>Test result: %s',
        Gene_Id, Start , DUP_test
          )
    )
  ) +
  facet_wrap(~DUP , ncol=1 , scales = "free_y", labeller = as_labeller(category_conversion_2) ) +
  scale_y_continuous(breaks = c(0,1) , limits=c(0,1.1) , labels = c("False" , "True")) +
  scale_colour_manual(values = brewer.pal(8, "Dark2")[3:6] ) +
  labs(x = "Sequence position (bp)", y = "Test result" , color = "")

p4 <- p4 + theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
                 axis.title.x = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.title.y = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.text.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
                 axis.text.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
                 axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
                 plot.background = element_blank(),
                 panel.background = element_rect(fill = "white", size = 0.5, linetype = "solid"),
                 panel.grid.major.y = element_line(color="#777777", size=0.2, linetype="solid"),
                 panel.grid.major.x = element_blank(),
                 panel.grid.minor.y = element_blank(),
                 panel.grid.minor.x = element_blank(),
                 plot.margin=unit(c(.7,.7,.7,.7),"mm"),
                 legend.position="right",
                 panel.spacing = unit(.8, "lines") ,
                 strip.background = element_rect(fill = "#999999" ),
                 strip.text.x = element_text(margin = margin(.3, 0, .1, 0, "lines") , color = "white ")
    )

pl4 <- ggplotly(p4, tooltip = "text") %>% layout(annotations = b , legend=list(title="") , xaxis = list(title=list( text="Sequence position (bp)", font = list(size=13) ) ) )

```


```{r}
p7 <- ggplot(structure_trace_h1) +
  geom_rect(
    aes(
      xmin = Start,
      xmax = Stop,
      ymin = 0 ,
      ymax = 1 ,
      #colour = Seq_Id,
      fill = Seq_Id,
      group = 3,
      text = sprintf(
        'Component sequence ID: %s<br>Component region: %s-%s<br>Orientation: %s',
        Seq_Id, Start , Stop, orientation
          )
      ),
       ) +
  labs(colour = "" , fill="") +
  scale_y_continuous(breaks = c( 0 ,1)) +
  scale_x_continuous(limits = c(0, hap1Len)) +
  labs(x = "Sequence position (bp)", y = "")

p7 <- p7 +
          theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
                 axis.title.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.title.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm")),
                 axis.text.x = element_text(color="black", size=6, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
                 axis.text.y = element_text(color="black", size=6, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
                 axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
                 plot.background = element_blank(),
                 panel.background = element_blank(),
                 panel.grid.major.y = element_line(color="gray80", size=0.2, linetype="solid") ,
                 panel.grid.major.x = element_line(color="gray80", size=0.2, linetype="solid"),
                 panel.grid.minor.y = element_blank(),
                 panel.grid.minor.x = element_blank(),
                 plot.margin=unit(c(.7,.7,.7,.7),"mm"),
               legend.position="right",
               legend.margin = margin(10,10,10,10) ,
               legend.box.margin = margin(10,10,10,10) ,
            )

pl7  <- ggplotly(p7, tooltip = "text" ) %>% layout(annotations = e)
```


```{r}
p8 <- ggplot(legacy_trace_h1) +
  geom_rect(
    aes(
      xmin = Start,
      xmax = Stop,
      ymin = 0 ,
      ymax = 1 ,
      #colour = Seq_Id,
      fill = Seq_Id,
      group = 3,
      text = sprintf(
        'Component sequence ID: %s<br>Component region: %s-%s<br>Orientation: %s',
        Seq_Id, Start , Stop, orientation
          )
      ),
       ) +
  labs(colour = "" , fill="") +
  scale_y_continuous(breaks = c( 0 ,1)) +
  scale_x_continuous(limits = c(0, hap1Len)) +
  labs(x = "Sequence position (bp)", y = "")

p8 <- p8 +
          theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
                 axis.title.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.title.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm")),
                 axis.text.x = element_text(color="black", size=6, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
                 axis.text.y = element_text(color="black", size=6, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
                 axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
                 plot.background = element_blank(),
                 panel.background = element_blank(),
                 panel.grid.major.y = element_line(color="gray80", size=0.2, linetype="solid") ,
                 panel.grid.major.x = element_line(color="gray80", size=0.2, linetype="solid"),
                 panel.grid.minor.y = element_blank(),
                 panel.grid.minor.x = element_blank(),
                 plot.margin=unit(c(.7,.7,.7,.7),"mm"),
               legend.position="right",
               legend.margin = margin(10,10,10,10) ,
               legend.box.margin = margin(10,10,10,10) ,
            )

pl8  <- ggplotly(p8, tooltip = "text" ) %>% layout(annotations = f)
```



```{r}
p11 <- ggplot(structure_trace_h2) +
  geom_rect(
    aes(
      xmin = Start,
      xmax = Stop,
      ymin = 0 ,
      ymax = 1 ,
      #colour = Seq_Id,
      fill = Seq_Id,
      group = 3,
      text = sprintf(
        'Component sequence ID: %s<br>Component region: %s-%s<br>Orientation: %s',
        Seq_Id, Start , Stop, orientation
          )
      ),
       ) +
  labs(colour = "" , fill="") +
  scale_y_continuous(breaks = c( 0 ,1)) +
  scale_x_continuous(limits = c(0, hap2Len)) +
  labs(x = "Sequence position (bp)", y = "")

p11 <- p11 +
          theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
                 axis.title.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.title.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm")),
                 axis.text.x = element_text(color="black", size=6, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
                 axis.text.y = element_text(color="black", size=6, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
                 axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
                 plot.background = element_blank(),
                 panel.background = element_blank(),
                 panel.grid.major.y = element_line(color="gray80", size=0.2, linetype="solid") ,
                 panel.grid.major.x = element_line(color="gray80", size=0.2, linetype="solid"),
                 panel.grid.minor.y = element_blank(),
                 panel.grid.minor.x = element_blank(),
                 plot.margin=unit(c(.7,.7,.7,.7),"mm"),
               legend.position="right",
               legend.margin = margin(10,10,10,10) ,
               legend.box.margin = margin(10,10,10,10) ,
            )

pl11  <- ggplotly(p11, tooltip = "text" ) %>% layout(annotations = e)
```

```{r}
p12 <- ggplot(legacy_trace_h2) +
  geom_rect(
    aes(
      xmin = Start,
      xmax = Stop,
      ymin = 0 ,
      ymax = 1 ,
      #colour = Seq_Id,
      fill = Seq_Id,
      group = 3,
      text = sprintf(
        'Component sequence ID: %s<br>Component region: %s-%s<br>Orientation: %s',
        Seq_Id, Start , Stop, orientation
          )
      ),
       ) +
  labs(colour = "" , fill="") +
  scale_y_continuous(breaks = c( 0 ,1)) +
  scale_x_continuous(limits = c(0, hap2Len)) +
  labs(x = "Sequence position (bp)", y = "")

p12 <- p12 +
          theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
                 axis.title.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.title.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm")),
                 axis.text.x = element_text(color="black", size=6, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
                 axis.text.y = element_text(color="black", size=6, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
                 axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
                 plot.background = element_blank(),
                 panel.background = element_blank(),
                 panel.grid.major.y = element_line(color="gray80", size=0.2, linetype="solid") ,
                 panel.grid.major.x = element_line(color="gray80", size=0.2, linetype="solid"),
                 panel.grid.minor.y = element_blank(),
                 panel.grid.minor.x = element_blank(),
                 plot.margin=unit(c(.7,.7,.7,.7),"mm"),
               legend.position="right",
               legend.margin = margin(10,10,10,10) ,
               legend.box.margin = margin(10,10,10,10) ,
            )

pl12  <- ggplotly(p12, tooltip = "text" ) %>% layout(annotations = f)
```

```{r}

p13 <- ggplot(diploid_trace_ref,
    aes(
      x = Start,
      y = count,
      color = HAP,
      group = 1,
      text = sprintf(
        'Gene ID: %s<br>Gene start: %s<br>%s: %s',
        Gene_Id, Start , HAP , count
          )
    )
  ) +
  geom_step() +
  scale_colour_manual(values = brewer.pal(8, "Dark2")[1:2] , labels = c("Hap 1 counts" , "Hap 1 counts") ) +
  scale_y_continuous(breaks = seq(min(diploid_trace11$count), max(diploid_trace11$count), 1)) +
  labs(x = "Sequence position (bp)", y = "Gene alignments counts" , color = "")

p13 <- p13 +
          theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
                 axis.title.x = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.title.y = element_text(color="black", size=10, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm")),
                 axis.text.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
                 axis.text.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
                 axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
                 plot.background = element_blank(),
                 panel.background = element_blank(),
                 panel.grid.major.y = element_blank(),
                 panel.grid.major.x = element_blank(),
                 panel.grid.minor.y = element_blank(),
                 panel.grid.minor.x = element_blank(),
                 plot.margin=unit(c(.7,.7,.7,.7),"mm"),
                 legend.position="right"
            )

pl13 <- ggplotly(p13, tooltip = "text") %>% layout(annotations = a , legend=list(title="") , xaxis = list(title="Sequence position (bp)") )

```



```{r}
p16 <- ggplot(structure_trace_r) +
  geom_rect(
    aes(
      xmin = Start,
      xmax = Stop,
      ymin = 0 ,
      ymax = 1 ,
      #colour = Seq_Id,
      fill = Seq_Id,
      group = 3,
      text = sprintf(
        'Component sequence ID: %s<br>Component region: %s-%s<br>Orientation: %s',
        Seq_Id, Start , Stop, orientation
          )
      ),
       ) +
  labs(colour = "" , fill="") +
  scale_y_continuous(breaks = c( 0 ,1)) +
  scale_x_continuous(limits = c(0, refLen)) +
  labs(x = "Sequence position (bp)", y = "")

p16 <- p16 +
          theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
                 axis.title.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.title.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm")),
                 axis.text.x = element_text(color="black", size=6, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
                 axis.text.y = element_text(color="black", size=6, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
                 axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
                 plot.background = element_blank(),
                 panel.background = element_blank(),
                 panel.grid.major.y = element_line(color="gray80", size=0.2, linetype="solid") ,
                 panel.grid.major.x = element_line(color="gray80", size=0.2, linetype="solid"),
                 panel.grid.minor.y = element_blank(),
                 panel.grid.minor.x = element_blank(),
                 plot.margin=unit(c(.7,.7,.7,.7),"mm"),
               legend.position="right",
               legend.margin = margin(10,10,10,10) ,
               legend.box.margin = margin(10,10,10,10) ,
            )

pl16  <- ggplotly(p16, tooltip = "text" ) %>% layout(annotations = e)
```

```{r}
p17 <- ggplot(legacy_trace_r) +
  geom_rect(
    aes(
      xmin = Start,
      xmax = Stop,
      ymin = 0 ,
      ymax = 1 ,
      #colour = Seq_Id,
      fill = Seq_Id,
      group = 3,
      text = sprintf(
        'Component sequence ID: %s<br>Component region: %s-%s<br>Orientation: %s',
        Seq_Id, Start , Stop, orientation
          )
      ),
       ) +
  labs(colour = "" , fill="") +
  scale_y_continuous(breaks = c( 0 ,1)) +
  scale_x_continuous(limits = c(0, refLen)) +
  labs(x = "Sequence position (bp)", y = "")

p17 <- p17 +
          theme(plot.title = element_text(color="black", size=10, face="plain", margin=margin(t=1,r=0,b=1,l=0,"mm")),
                 axis.title.x = element_text(color="black", size=8, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm")),
                 axis.title.y = element_text(color="black", size=8, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm")),
                 axis.text.x = element_text(color="black", size=6, face="plain", margin=margin(t=1,r=0,b=0,l=0,"mm"), hjust = 0.5),
                 axis.text.y = element_text(color="black", size=6, face="plain", margin=margin(t=0,r=1,b=0,l=0,"mm"), vjust = 0.5),
                 axis.line.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.line.y = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.x = element_line(color="black", size=0.5, linetype="solid"),
                 axis.ticks.y = element_line(color="black", size=0.5, linetype="solid"),
                 plot.background = element_blank(),
                 panel.background = element_blank(),
                 panel.grid.major.y = element_line(color="gray80", size=0.2, linetype="solid") ,
                 panel.grid.major.x = element_line(color="gray80", size=0.2, linetype="solid"),
                 panel.grid.minor.y = element_blank(),
                 panel.grid.minor.x = element_blank(),
                 plot.margin=unit(c(.7,.7,.7,.7),"mm"),
               legend.position="right",
               legend.margin = margin(10,10,10,10) ,
               legend.box.margin = margin(10,10,10,10) ,
            )

pl17  <- ggplotly(p17, tooltip = "text" ) %>% layout(annotations = f)
```


Dotplot
=====
```{r}
### Dotplot of alignment results
subplot( pl0 , pl16 , pl17 , pl13 , nrows = 4, shareX = T , titleY = T , titleX=T , heights = c(0.55, 0.1 , 0.1 ,0.25))
```

Gene counts : Hap1 genes
=====
### Hap1 genes alignment count
```{r}
subplot(pl7 , pl8 , pl1, style(pl2 , showlegend = F) , nrows = 4, shareX = T , titleY = T , titleX=T , heights = c(0.08, 0.1 , 0.22 ,0.45))
```


Gene counts : Hap2 genes
=====
### Hap2 genes alignment count
```{r}
subplot(pl11 , pl12 , pl3, style(pl4 , showlegend = F) , nrows = 4, shareX = T , titleY = T , titleX=T , heights = c(0.08, 0.1 , 0.22 ,0.45))
```
